From 6cae766b5baf7b46ec9cb275c0f957bc1ae1db63 Mon Sep 17 00:00:00 2001
From: Vaibhav Bedia <vaibhav.bedia@ti.com>
Date: Sat, 30 Jun 2012 22:10:08 +0800
Subject: [PATCH 068/609] ARM: OMAP: AM33XX: Don't register PRCM chain handler

Original commit: dcd92aca6a2cbc009d089becb37c77d933ae738f

  From git://arago-project.org/git/projects/linux-am33x.git
  And ti-sdk-am335x-evm-05.05.00.00-Linux-x86-Install image

AM33XX does not support PRCM interrupts for IO/WKUP events.
Since the current PRCM chain handler expects only these two
events, don't register it for AM33XX.

We might revisit this once the PRCM chain handler code is
enhanced to take care of other events like DPLL needing to be
recalibrated.

Note: This fixes the following crash in USB during kernel boot

[    2.131963] Unhandled fault: external abort on non-linefetch
(0x1008) at 0xc880a460
[    2.139960] Internal error: : 1008 [#1]
[    2.143954] Modules linked in:
[    2.147143] CPU: 0    Not tainted  (3.2.0-rc6-12008-g363a774)
[    2.154057] PC is at __musb_readb+0xc/0x14
[    2.158329] LR is at otg_timer+0x2c/0x13c
...

This was due to the PRCM_IRQENABLE register for OMAP3 being at
the same offset as USB_CLKCTRL on AM33XX.

Signed-off-by: Vaibhav Bedia <vaibhav.bedia@ti.com>
---
 arch/arm/mach-omap2/prm2xxx_3xxx.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-omap2/prm2xxx_3xxx.c b/arch/arm/mach-omap2/prm2xxx_3xxx.c
index 21cb740..1f7c0cb 100644
--- a/arch/arm/mach-omap2/prm2xxx_3xxx.c
+++ b/arch/arm/mach-omap2/prm2xxx_3xxx.c
@@ -306,6 +306,8 @@ static int __init omap3xxx_prcm_init(void)
 {
 	int ret = 0;
 
+	if (cpu_is_am33xx())
+		return 0;
 	if (cpu_is_omap34xx()) {
 		ret = omap_prcm_register_chain_handler(&omap3_prcm_irq_setup);
 		if (!ret)
-- 
1.7.5.4

